src/controller/proxy/local.go

```diff
@@ -0,0 +1,182 @@
+}
+
+// newLocalHelper create the localInterface
+func newLocalHelper() localInterface {
+	l := &localHelper{artifactCtl: artifact.Ctl}
+	l.init()
+	return l
+}
+
+func (l *localHelper) BlobExist(ctx context.Context, art lib.ArtifactInfo) (bool, error) {
+	return l.registry.BlobExist(art.Repository, art.Digest)
+}
+
+func (l *localHelper) init() {
+	if l.registry != nil {
+		return
+	}
+	log.Debugf("core url:%s, local core url: %v", config.GetCoreURL(), config.LocalCoreURL())
+	// the traffic is internal only
+	registryURL := config.LocalCoreURL()
+	authorizer := comHttpAuth.NewSecretAuthorizer(config.ProxyServiceSecret)
+	l.registry = registry.NewClientWithAuthorizer(registryURL, authorizer, true)
+}
+
+func (l *localHelper) PushBlob(localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error {
+	log.Debugf("Put blob to local registry, localRepo:%v, digest: %v", localRepo, desc.Digest)
+	err := l.registry.PushBlob(localRepo, string(desc.Digest), desc.Size, bReader)
+	return err
+}
+
+func (l *localHelper) PushManifest(repo string, tag string, manifest distribution.Manifest) error {
+	// Make sure there is only one go routing to push current artName to local repo
+	if len(tag) == 0 {
+		// when push a manifest list, the tag is empty, for example: busybox
+		// if tag is empty, set to latest
+		tag = "latest"
+	}
+	artName := repo + ":" + tag
+	if !inflightChecker.addRequest(artName) {
+		return nil
+	}
+	defer inflightChecker.removeRequest(artName)
+
+	mediaType, payload, err := manifest.Payload()
+	if err != nil {
+		return err
+	}
+	_, err = l.registry.PushManifest(repo, tag, mediaType, payload)
+	return err
+}
+
+// DeleteManifest cleanup delete tag from local cache
+func (l *localHelper) DeleteManifest(repo, ref string) {
+	log.Debug("Remove tag from repo if it is exist")
+	if err := l.registry.DeleteManifest(repo, ref); err != nil {
+		// sometimes user pull a non-exist image
+		log.Debugf("failed to remove artifact, error %v", err)
+	}
+}
+
+// updateManifestList -- Trim the manifest list, make sure all depend manifests are ready
+func (l *localHelper) updateManifestList(ctx context.Context, repo string, manifest distribution.Manifest) (distribution.Manifest, error) {
+	switch v := manifest.(type) {
+	case *manifestlist.DeserializedManifestList:
+		existMans := make([]manifestlist.ManifestDescriptor, 0)
+		for _, m := range v.Manifests {
+			art := lib.ArtifactInfo{Repository: repo, Digest: string(m.Digest)}
+			exist, err := l.ManifestExist(ctx, art)
+			if err != nil {
+				continue
+			}
+			if exist {
+				existMans = append(existMans, m)
+			}
+		}
+		if len(existMans) > 0 {
+			// Avoid empty manifest in the manifest list
+			return manifestlist.FromDescriptors(existMans)
+		}
+	}
+	return manifest, nil
+}
+
+func (l *localHelper) PushManifestList(ctx context.Context, repo string, tag string, man distribution.Manifest) error {
```

Comment:
This may not block merging this PR, but the way to handle manifest list can be refined, could you open a new issue to track that work?

Reply:
Updated to wait and check strategy